File: backend/services/leverage/tariff.py
# services/leverage/tariff.py
"""
Тарифный модуль: цена за риск плечевой позиции.

  тариф = стоимость капитала + гэп-премия(λ·J·запас) + маржа платформы

Все три — в долях от маржи игрока. λ/J приходят через LambdaJSource (шов).
"""
from __future__ import annotations

from dataclasses import dataclass

from knockout import Position
from lambda_source import LambdaJSource


@dataclass(frozen=True)
class TariffConfig:
    capital_rate: float = 0.02      # стоимость капитала: доля от обязательства пула
    platform_fee: float = 0.015     # маржа платформы: доля от маржи
    gap_buffer: float = 1.5         # запас поверх λ·J (неопределённость)
    tick_value: float = 0.01        # $ за 1 тик (1 тик = $0.01)


@dataclass(frozen=True)
class TariffBreakdown:
    capital_cost: float   # $ за замороженный капитал
    gap_premium: float    # $ ожидаемого убытка от прыжков + запас
    platform_cost: float  # $ заработок платформы
    total: float          # $ итого с игрока
    total_pct: float      # % от маржи

    def as_dict(self) -> dict:
        return {
            "capital_cost": round(self.capital_cost, 4),
            "gap_premium": round(self.gap_premium, 4),
            "platform_cost": round(self.platform_cost, 4),
            "total": round(self.total, 4),
            "total_pct": round(self.total_pct, 2),
        }


def barrier_ticks(pos: Position) -> float:
    """Расстояние от входа до нокаута в тиках — это d для калибровки λ/J."""
    return abs(pos.entry - pos.knockout)


def price_position(pos: Position, source: LambdaJSource,
                   cfg: TariffConfig = TariffConfig()) -> TariffBreakdown:
    """Собрать тариф из трёх слагаемых для конкретной позиции."""
    # 1) стоимость капитала: пул морозит обязательство (max_payout)
    capital_cost = pos.max_payout * cfg.capital_rate

    # 2) гэп-премия: λ·J на нокаут-барьере, в $, с запасом.
    #    λ·J — ожидаемый перелёт в тиках на позицию -> в $ через размер и tick_value.
    d = barrier_ticks(pos)
    lam, j = source.lookup(d)
    expected_overshoot_ticks = lam * j            # ожидаемый перелёт, тиков
    gap_premium = (expected_overshoot_ticks * cfg.tick_value
                   * pos.size * cfg.gap_buffer)   # -> $, масштаб позиции, с запасом

    # 3) маржа платформы: доля от маржи игрока
    platform_cost = pos.margin * cfg.platform_fee

    total = capital_cost + gap_premium + platform_cost
    total_pct = total / pos.margin * 100.0

    return TariffBreakdown(
        capital_cost=capital_cost, gap_premium=gap_premium,
        platform_cost=platform_cost, total=total, total_pct=total_pct,
    )

File: src/lib/pricing/tariff.ts
/**
 * Тарифный модуль: цена за риск плечевой позиции.
 * Порт services/leverage/tariff.py.
 *
 *   тариф = стоимость капитала + гэп-премия(λ·J·запас) + маржа платформы
 *
 * λ/J приходят через `LambdaJSource` (шов), поэтому подмена калибровки
 * не трогает формулу.
 */

import type { Position } from "./knockout";
import type { LambdaJSource } from "./lambda-source";

export interface TariffConfig {
  /** Стоимость капитала: доля от обязательства пула. */
  capitalRate: number;
  /** Маржа платформы: доля от маржи игрока. */
  platformFee: number;
  /** Запас поверх λ·J — плата за неопределённость калибровки. */
  gapBuffer: number;
  /** $ за 1 тик. */
  tickValue: number;
}

export const DEFAULT_TARIFF: TariffConfig = {
  capitalRate: 0.02,
  platformFee: 0.015,
  gapBuffer: 1.5,
  tickValue: 0.01,
};

export interface TariffBreakdown {
  /** $ за замороженный капитал пула. */
  capitalCost: number;
  /** $ ожидаемого убытка от прыжков цены + запас. */
  gapPremium: number;
  /** $ заработок платформы. */
  platformCost: number;
  /** $ итого с игрока. */
  total: number;
  /** % от маржи. */
  totalPct: number;
}

/** Пустой разбор — для котировки с ошибкой, чтобы в UI не было NaN. */
export const EMPTY_TARIFF: TariffBreakdown = {
  capitalCost: 0,
  gapPremium: 0,
  platformCost: 0,
  total: 0,
  totalPct: 0,
};

/** Расстояние от входа до нокаута в тиках — это d для калибровки λ/J. */
export function barrierTicks(pos: Position): number {
  return Math.abs(pos.entry - pos.knockout);
}

/** Собрать тариф из трёх слагаемых для конкретной позиции. */
export function pricePosition(
  pos: Position,
  source: LambdaJSource,
  cfg: TariffConfig = DEFAULT_TARIFF,
): TariffBreakdown {
  // 1) стоимость капитала: пул морозит обязательство (maxPayout)
  const capitalCost = pos.maxPayout * cfg.capitalRate;

  // 2) гэп-премия: λ·J на нокаут-барьере — ожидаемый перелёт в тиках;
  //    переводим в $ через размер позиции и цену тика, сверху запас.
  const { lambda, j } = source.lookup(barrierTicks(pos));
  const expectedOvershootTicks = lambda * j;
  const gapPremium = expectedOvershootTicks * cfg.tickValue * pos.size * cfg.gapBuffer;

  // 3) маржа платформы: доля от маржи игрока
  const platformCost = pos.margin * cfg.platformFee;

  const total = capitalCost + gapPremium + platformCost;

  return {
    capitalCost,
    gapPremium,
    platformCost,
    total,
    totalPct: (total / pos.margin) * 100,
  };
}

/** Округление под показ — аналог их `TariffBreakdown.as_dict()`. */
export function roundTariff(b: TariffBreakdown): TariffBreakdown {
  const r = (v: number, digits: number): number => {
    const k = 10 ** digits;
    return Math.round(v * k) / k;
  };
  return {
    capitalCost: r(b.capitalCost, 4),
    gapPremium: r(b.gapPremium, 4),
    platformCost: r(b.platformCost, 4),
    total: r(b.total, 4),
    totalPct: r(b.totalPct, 2),
  };
}
